### PR TITLE
[sparse] Add padding for dense matrices in semi-structured sparse

### PR DESCRIPTION
**Pull request by jcaip**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111049
* __->__ #110583

Summary:

Currently we have shape constraints in semi-structured sparsity for both
CUTLASS and cuSPARSELt

These shape constraints unfortunately apply to both the dense and sparse
matrices in sparsedense matmul.

This PR adds in support for calling `F.pad` in order to pad dense
matrices to the right size with zeros and then pull out the
corresponding rows from the resultant result matrix.

We also throw a warning in this case.
The tests have also been updated to take in a dense_input_shape
parameter.

Test Plan:
```
python test/test_sparse_semi_structured.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: torch/sparse/semi_structured.py [hunk 1]
```diff
@@ -232,6 +232,33 @@ def __repr__(self) -> str:  # type: ignore[override]
...
+        assert original_tensor.dim() == 2
+
+        # check shape
+        m, n = original_tensor.shape
+        min_rows = _DTYPE_TO_SEMI_STRUCTURED_SPARSE_CONFIG[original_tensor.dtype].min_rows
+        min_cols = _DTYPE_TO_SEMI_STRUCTURED_SPARSE_CONFIG[original_tensor.dtype].min_cols
+        to_pad_m = -m % min_rows if m < min_rows or m % min_rows else 0
+        to_pad_n = -n % min_cols if n < min_cols or n % min_rows else 0
+        if to_pad_m or to_pad_n:
+            warnings.warn(
```
Comment (cpuhrsch) [inline torch/sparse/semi_structured.py:250]: Will this warn on every invocation?

File: torch/sparse/semi_structured.py [hunk 2]
```diff
@@ -232,6 +232,33 @@ def __repr__(self) -> str:  # type: ignore[override]
...
+                (
+                    "Attempting to do matmul with a dense tensor that does not meet shape requirements."
+                    "Padding dense input tensor of shape ({m}, {n}) to ({m+to_pad_m}, {n+to_pad_n}."
+                ),
+                UserWarning,
+            )
+            return torch.nn.functional.pad(original_tensor, (0, to_pad_n, 0, to_pad_m))
```
Comment (cpuhrsch) [inline torch/sparse/semi_structured.py:257]: I think (hope) F.pad is smart enough to do nothing when `to_pad` is 0. 

Also, can you try this code with torch.compile to see how the compiler handles the padding?

File: torch/sparse/semi_structured.py (hunk 1 above)
Comment (jcaip) [reply torch/sparse/semi_structured.py:250]: Yes, we can change this to on first invocation.

File: torch/sparse/semi_structured.py (hunk 2 above)
Comment (jcaip) [reply torch/sparse/semi_structured.py:257]: Yes, I will run our benchmarks again too to make sure there is no perf regression. 

For torch.compile we don't have brians subclassing support yet. 
Do you want me to just verify that we can fuse padding with a test? 

I can also just open a new branch with the same padding changes in saf, if you want to try this e2e

Comment (cpuhrsch) [reply torch/sparse/semi_structured.py:257]: @jcaip I think Brian's work actually just recently landed (at least I thought it did)

**Review by alexsamardzic — APPROVED** [review]
LGTM.

File: torch/sparse/semi_structured.py (hunk 2 above)
Comment (jcaip) [reply torch/sparse/semi_structured.py:257]: Oh perfect, I'll try that then

File: torch/sparse/semi_structured.py (hunk 1 above)
Comment (jcaip) [reply torch/sparse/semi_structured.py:250]: cc @cpuhrsch 

Actually first invocation might be tricky because this is really a per-matmul warning. So we can't have a warning_shown class variable flag.

And since we create a new instance of the subclass we cannot store if the warning is shown in an instance variable. I will probably add a flag to disable all padding warnings instead. Perhaps we should make a `padding_mode` flag that needs to be set before we automatically do padding.

**Comment by jcaip** [comment]
I see no perf difference for padding no-op on nvidia-bert benchmarks


with padding:
```
  test_function     m     k      n          dtype  backend  sparse_latency (ms)  dense_latency (ms)  speedup (d/s)  correct  contiguous
0        linear  3072  1024  16384  torch.float16  cutlass             0.310758            0.447545       1.440169     True       False
1        linear  4096  1024  16384  torch.float16  cutlass             0.410349            0.586863       1.430155     True       False
2        linear  1024  1024  16384  torch.float16  cutlass             0.112575            0.178476       1.585390     True       False
3        linear  1024  4096  16384  torch.float16  cutlass             0.373492            0.560101       1.499633     True       False

```

no padding:
```
  test_function     m     k      n          dtype  backend  sparse_latency (ms)  dense_latency (ms)  speedup (d/s)  correct  contiguous
0        linear  3072  1024  16384  torch.float16  cutlass             0.311111            0.448242       1.440778     True       False
1        linear  4096  1024  16384  torch.float16  cutlass             0.411172            0.593480       1.443387     True       False
2        linear  1024  1024  16384  torch.float16  cutlass             0.113998            0.178642       1.567061     True       False
3        linear  1024  4096  16384  torch.float16  cutlass             0.375681            0.560720       1.492543     True       False
```

File: torch/sparse/semi_structured.py (hunk 1 above)
Comment (cpuhrsch) [reply torch/sparse/semi_structured.py:250]: I'd say we remove the warning for now. If someone can't use the 2:4 format because their inputs don't align they have to pad. If it's slower because of padding, then they'll need to adjust inputs. But at least they can use it and try.


But a warning on every invocation can cause a lot of issues and slowdowns (I think even graph breaks maybe?)

**Review by cpuhrsch — APPROVED** [review]
